Misc fixes for EntitySnapshot

ImMorpheus opened the pull request:
Fix `EntitySnapshot.builder().build().toContainer();` NPE
Fix `EntitySnapshot.builder().build(EntitySnapshot.builder().build().toContainer());` `Optional#empty()`
Fix tests
Fix EntitySnapshot hash-based collections (missing hashcode/equals)

ImMorpheus committed:
Implement equals and hashcode for EntitySnapshot

gabizou commented:
EntitySnapshots are not supposed to be treated like EntityArchetypes in that they need AT LEAST valid information from an Entity to be created. Much like BlockSnapshots in a sense for TileEntity information. 

The issue with changing this is that you now make it acceptable to abuse snapshots to restore or create them as necessary, without really differing them from EntityArchetypes.

ImMorpheus commented:
> EntitySnapshots are not supposed to be treated like EntityArchetypes in that they need AT LEAST valid information from an Entity to be created

That's not the case at all. There's nothing stopping you from doing `EntitySnapshot.builder().build()`
This means the current system is broken. Working on it.

ImMorpheus committed:
Avoid building broken EntitySnapshots

EntitySnapshots are not supposed to be treated like EntityArchetypes in that they need AT LEAST valid information from an Entity to be created

ImMorpheus committed:
Fix tests

ImMorpheus committed:
try a bit harder

ImMorpheus committed:
Fix tests again
